debezium-server-redis/src/main/java/io/debezium/server/redis/RedisMemoryThreshold.java

```diff
@@ -26,56 +25,78 @@ public class RedisMemoryThreshold {
     private static final String INFO_MEMORY = "memory";
     private static final String INFO_MEMORY_SECTION_MAXMEMORY = "maxmemory";
     private static final String INFO_MEMORY_SECTION_USEDMEMORY = "used_memory";
-
-    private static final Supplier<Boolean> MEMORY_OK = () -> true;
+    private static long accumulatedMemory = 0L;
+    private static long previouslyUsedMemory = 0L;
+    private static long totalProcessed = 0;
 
     private RedisClient client;
-
-    private int memoryThreshold;
-
-    private long memoryLimit;
-
-    private Supplier<Boolean> isMemoryOk;
+    private long memoryLimit = 0;
+    private long maximumMemory = 0;
 
     public RedisMemoryThreshold(RedisClient client, RedisStreamChangeConsumerConfig config) {
         this.client = client;
-        this.memoryThreshold = config.getMemoryThreshold();
         this.memoryLimit = 1024L * 1024 * config.getMemoryLimitMb();
-        if (memoryThreshold == 0 || memoryTuple(memoryLimit) == null) {
-            disable();
-        }
-        else {
-            this.isMemoryOk = () -> isMemoryOk();
-        }
     }
 
-    public boolean check() {
-        return isMemoryOk.get();
+    /**
+     * @param client
+     */
+    public void setRedisClient(RedisClient client) {
+        this.client = client;
     }
 
-    private boolean isMemoryOk() {
-        Tuple2<Long, Long> memoryTuple = memoryTuple(memoryLimit);
-        if (memoryTuple == null) {
-            disable();
+    /**
+     * @param extraMemory   - Estimated size of a single record.
+     * @param bufferSize    - Number of records in a batch.
+     * @param bufferFillRate - Rate in which memory can be filled.
+     * @return
+     */
+    public boolean checkMemory(long extraMemory, int bufferSize, int bufferFillRate) {
+        Tuple2<Long, Long> memoryTuple = memoryTuple();
+
+        if (maximumMemory == 0) {
+            totalProcessed += bufferSize;
+            LOGGER.debug("Total Processed Records: {}", totalProcessed);
             return true;
         }
-        long maxMemory = memoryTuple.getItem2();
-        if (maxMemory > 0) {
-            long usedMemory = memoryTuple.getItem1();
-            long percentage = usedMemory * 100 / maxMemory;
-            if (percentage >= memoryThreshold) {
-                LOGGER.info("Sink memory threshold percentage was reached. Will retry; " +
-                        "(current: {}%, configured: {}%, used_memory: {}, maxmemory: {}).",
-                        percentage, memoryThreshold,
-                        usedMemory, maxMemory);
-                return false;
-            }
+
+        maximumMemory = memoryTuple.getItem2();
+
+        long extimatedBatchSize = extraMemory * bufferFillRate;
+        long usedMemory = memoryTuple.getItem1();
+        long prevAccumulatedMemory = accumulatedMemory;
+        long diff = usedMemory - previouslyUsedMemory;
+        if (diff == 0L) {
```

Comment:
I don't understand reason for this. You are doing the query every time. What is the reasoning about caching of used memory in `previouslyUsedMemory`? You could always operate on the most resent used memory. Also in that case `accumulatedMemory` is probably also unnecessary.

Reply:
Explained in a chat